tiledb/sm/rest/rest_client.h

```diff
@@ -27,7 +27,143 @@
+ * the example.
+ *
+ *   - `Array::Array` is created with a URI. That URI can be either of two
+ *     possibilities. The member variable `remote_` is initialized accordingly.
+ *       - Direct URI to an object store or file system. (`remote_ == false`)
+ *       - REST URI that begins with `tiledb:`. (`remote_ == true`)
+ *   - Calling `Array::open` implements two different ways of opening the file,
+ *     depending on whether it's remote or not.
+ *       - Direct (non-remote) opening does not use `RestClient` in any way.
+ *         This operation does not create any cycles.
+ *       - Remote opening always uses `RestClient`, calling
+ *         `get_array_schema_from_rest` and `post_array_from_rest`.
+ *   - As a particular example, `post_array_from_rest` calls
+ *     `array_deserialize`. (All `RestClient` operations use capnp
+ *     serialization functions analogously.)
+ *   - `array_deserialize` calls the `Array` constructor.
+ *
+ * Compacting and summarizing the call sequence:
+ *   - `Array` depends on `RestClient` because of `Array::open` uses
+ *     `RestClient::post_array_from_rest`.
+ *   - `RestClient` depends on `Array` because of
+ *     `RestClient::post_array_from_rest` uses `Array::Array`.
+ *
+ * The root cause of the cycle is that the `Array` constructor commingles both
+ * direct and remote URIs into the same argument. As long as this persists,
+ * there's a necessary cycle. This example is a specific example of a widespread
+ * and generic pattern in the use of `RestClient`.
+ *
+ * The minimum requirement is that there be two separate constructors and that
+ * they be defined in different translation units (so that they don't end up in
+ * the same object file). These two constructors might both be on an existing
+ * class itself, possibly, or they might be on different classes. Regardless of
+ * the approach, the elimination of these cycles is a long term project.
+ *
+ * @section Dependency on TILEDB_SERIALIZATION
+ *
+ * The REST client has two different behaviors depending on the build
+ * configuration for serialization.
+ *   - If serialization is off, the REST client is a stub, throwing for all
+ *     remote operations.
+ *   - If serialization is on, the REST client performs remote operations
+ *     with a server.
+ *
+ * The earliest way this polymorphism was accomplished was with two different
+ * versions of the class, both named `class RestClient`, selected with the
+ * preprocessor. This was a distinctly C-style way of programming with a number
+ * of disadvantages. (All these are corollaries of the general maxim that The
+ * Preprocessor is Evil).
+ *   - Only a single configuration could be compiled in a build tree at a time.
+ *   - It was impossible to test both kinds of REST client with any single
+ *     configuration.
+ *   - All object libraries with any dependence on `class RestClient` came in
+ *     two different versions, but with the same name.
```

Comment:
I don't understand this.

Reply:
The object libraries do not contain anything in their name that says whether it's linked to a rest_client stub or the full version. Thus there are two versions of the artifact. You might think that they'd never be confused, but that would be wrong.

I left alternate approaches out of this documentation. They were not going to work well. The problem got quite thorny when considering how to link the main library through object libraries, and yet also have those object libraries work seamlessly with standalone unit tests.